cardano-cli/test/Test/CLI/TextEnvelope/Golden/Genesis/Create.hs

```diff
@@ -0,0 +1,132 @@
+  ( (^?)
+  , (^..)
+  , each
+  )
+import Prelude(String) 
+
+import qualified Control.Lens         as CL
+import qualified Data.Aeson           as J
+import qualified Data.Aeson.Lens      as J
+import qualified Data.HashMap.Strict  as HM
+import qualified Data.List            as L
+import qualified Data.Set             as S
+import qualified Data.Text            as T
+import qualified Data.Time.Clock      as DT
+import qualified Data.Time.Format     as DT
+import qualified Data.Vector          as DV
+import qualified System.IO            as IO
+import qualified System.IO.Temp       as IO
+
+import Hedgehog
+  ( Property
+  , forAll
+  , (===)
+  )
+
+import qualified Hedgehog       as H
+import qualified Hedgehog.Gen   as G
+import qualified Hedgehog.Range as R
+import qualified Test.OptParse  as OP
+
+{- HLINT ignore "Use camelCase" -}
+
+-- | Convert an object to an array of array.
+--
+-- For example: {"a": 1, "b": 2} -> [["a", 1], ["b", 2]]
+objectToArray :: J.Value -> J.Value
+objectToArray v = case v of
+  J.Object o -> J.Array (DV.fromList (fmap (\(ek, ev) -> J.Array (DV.fromList [J.String ek, ev])) (HM.toList o)))
+  a -> a
+
+-- | Assert the file contains the given number of occurences of the given string
+assertFileOccurences :: Int -> String -> FilePath -> H.PropertyT IO ()
+assertFileOccurences n s fp = do
+  signingKeyContents <- liftIO $ IO.readFile fp
+
+  length (filter (s `L.isInfixOf`) (L.lines signingKeyContents)) === n
+
+-- | Format the given time as an ISO 8601 date-time string
+formatIso8601 :: DT.UTCTime -> String
+formatIso8601 = DT.formatTime DT.defaultTimeLocale (DT.iso8601DateFormat (Just "%H:%M:%SZ"))
+
+-- | Return the supply value with the result of the supplied function as a tuple
+withSnd :: (a -> b) -> a -> (a, b)
+withSnd f a = (a, f a)
+
+golden_genesisCreate :: Property
+golden_genesisCreate = OP.propertyOnce $ do
+  tempDir <- liftIO $ IO.createTempDirectory "/Users/jky/wrk/iohk/cardano-node/cardano-cli/tmp" "test"
+  let genesisFile = tempDir <> "/genesis.json"
+  let cleanupPaths = [tempDir]
+  
+  fmtStartTime <- fmap formatIso8601 $ liftIO DT.getCurrentTime
+
+  (supply       , fmtSupply       ) <- fmap (withSnd show) $ forAll $ G.int (R.linear 10000000 4000000000)
+  (delegateCount, fmtDelegateCount) <- fmap (withSnd show) $ forAll $ G.int (R.linear 4 19)
+
+  -- Create the genesis json file and required keys
+  OP.execCardanoCLIParser cleanupPaths $
+    OP.evalCardanoCLIParser
+      [ "shelley","genesis","create"
+      , "--testnet-magic", "12"
+      , "--start-time", fmtStartTime
+      , "--supply", fmtSupply
+      , "--gen-genesis-keys", fmtDelegateCount
+      , "--gen-utxo-keys", fmtDelegateCount
+      , "--genesis-dir", tempDir
+      ]
+
+  OP.assertFilesExist [genesisFile]
+
+  genesisContents <- liftIO $ IO.readFile genesisFile
+
+  H.annotate genesisContents
+
+  actualSupply        <- forAll $ pure $ fromJust $ genesisContents ^? J.key "maxLovelaceSupply" . J._Integral
```

Comment:
Isn't `x <- forAll $ pure ...` the same as `let x = ...`

Reply:
Almost.

It's a thing I do to make `hedgehog` print the values when the test fails.  I can change them to `let`.

Reply:
Ah, ok, Nevermind. Leave it.